icu4c/source/i18n/unitsdata.cpp

```diff
@@ -94,8 +97,213 @@ class ConversionRateDataSink : public ResourceSink {
+ * TODO(hugovdm): implement the unit->category lookup (via "unitQuantities" in
+ * the units resource bundle).
+ * @param usage The usage for which formatting preferences is needed. If the
+ * given usage is not known, this function automatically falls back to "default"
+ * usage.
+ * @param region The region for which preferences are needed. If there are no
+ * region-specific preferences, this function automatically falls back to the
+ * "001" region (global).
+ * @param status The standard ICU error code output parameter. If an invalid
+ * category is given, status will be U_ILLEGAL_ARGUMENT_ERROR. If fallback to
+ * "default" or "001" didn't resolve, status will be U_MISSING_RESOURCE.
+ * @return The index into the metadata vector which represents the appropriate
+ * preferences. If appropriate preferences are not found, -1 is returned.
+ */
+int32_t getPreferenceIndex(const MaybeStackVector<UnitPreferenceMetadata> *metadata,
+                           const char *category, const char *usage, const char *region,
+                           UErrorCode &status) {
+    if (U_FAILURE(status)) { return -1; }
+    //  SearchStage stage, const char *category, const char *usage, const char *region) {
+    int32_t start = 0;
+    int32_t end = metadata->length();
+    bool foundCategory = false;
+    bool foundUsage = false;
+    int32_t checkpointStart = start;
+    int32_t checkpointEnd = end;
+    while (start < end) {
+        int32_t mid = (start + end) / 2;
+        int32_t cmp;
+        cmp = uprv_strcmp((*metadata)[mid]->category.data(), category);
+        if (cmp == 0) {
+            if (!foundCategory) {
+                foundCategory = true;
+                checkpointStart = start;
+                checkpointEnd = end;
+            }
+            cmp = uprv_strcmp((*metadata)[mid]->usage.data(), usage);
+            if (cmp == 0) {
+                if (!foundUsage) {
+                    foundUsage = true;
+                    checkpointStart = start;
+                    checkpointEnd = end;
+                }
+                cmp = uprv_strcmp((*metadata)[mid]->region.data(), region);
+                if (cmp == 0) {
+                    // We found a full metch.
+                    return mid;
+                }
+            }
+        }
+        if (cmp < 0) {
+            start = mid + 1;
+        } else {
+            U_ASSERT(cmp > 0);
+            end = mid;
+        }
+        if (start >= end) {
+            if (!foundCategory) {
+                // We don't do fallback categories - an invalid category was
+                // requested.
+                status = U_ILLEGAL_ARGUMENT_ERROR;
+                return false;
+            } else if (!foundUsage) {
+                if (uprv_strcmp(usage, "default") != 0) {
+                    // Try "default" usage. Every category should have at least
+                    // this usage.
+                    usage = "default";
+                    start = checkpointStart;
```

Comment:
This `checkpointStart` and `checkpointEnd` thing is cute, but binary searches are very fast.  If it's easy, you can keep this, but if changing to `operator<()` means you don't have the checkpoints any more, just go back to the full list on the second search.

Reply:
I think the cute checkpointing is more trouble than it's worth, if we're trying to have a "very simple and readable" binary search function.

Checkpointing needs plumbing of start/end state up to the top-level function, or a more side-effects-heavy (OO) style implementation instead of a "functional programming" style. Flattening out the structures again (not having this hierarchy of fucntions) looses the clean/simple binary search - then I doubt it would end up cleaner than my first-to-be-reviewed implementation. (It wasn't my first prototype. ;-) I already did a rewrite to try to make it as easily readable as possible.)

Reply:
OK.